api/health_check.proto

```diff
@@ -88,6 +88,14 @@ message HealthCheck {
   message RedisHealthCheck {
   }
 
+  // [#not-implemented-hide:] grpc.health.v1.Health-based healthcheck.
+  // See `GRPC doc <https://github.com/grpc/grpc/blob/master/doc/health-checking.md>`
+  // for details.
+  message GrpcHealthCheck {
+    // The name of the service to perform check on.
```

Comment:
If we go with service terminology, maybe be extra specific and clearly state something like: The semantics of service names are provided in https://github.com/grpc/grpc/blob/master/doc/health-checking.md. If we go with cluster terminology, please provide a similar mapping.

If we think of upstreams as always being part of an Envoy mesh, then cluster ~= service, and it would be good to stick to have consistent terminology as per @wora comment. OTOH, whenever we impedance match with the gRPC world, we end up having to bridge the cluster/service gap (I'm going this right now when integrating with the Google gRPC library implementation). In general, endpoints might be Google gRPC and not Envoy, at which point it makes more sense to use the "service name" terminology.

Reply:
Renamed to `service_name`. I've also added more elaborated comment on it and its connection with gRPC health protocol.